Cache DAGOpNode sort keys

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

In a typical circuit there are only certain numbers of qargs that are used regularly. The best examples are 1q gates on each qubit in the target, or 2q gates on each edge in the connectivity graph. As the qargs will be repeated many times in a typical circuit then for most operations the sort key attribute is also not-unique (as it's just composed of the positional indices of each qarg and carg). In very large circuit compilation we spend a non-insignificant amount of time constructing the sort key strings on nodes during the lifecycle of the DAGCircuit as it gets transformed. This commit caches the sort keys so that they're not rebuilt for every new DAGNode. This does add some extra overhead to the construction the first time a DAGNode is created because it has to do an additional qarg tuple construction and a dict lookup. But this is worth it because in most circutis the cache hit rate will be very high and we'll avoid the more expensive string construction for most gates in a circuit.

~There is also a potential shared state issue here if any qubit or clbit objects are present in >1 DAGCircuit and the relative positions of those bits differ between the circuits the sort key will not be the same as before because the earlier qubit combinations. This is because the cache is stored as a class attribute on the DAGOpNode class and will be reused between circuits. In practice this is unlikely to come up very frequently. But also it's impact is mitigated by these keys just being used for sorting and the sorting hasn't always been 100% deterministic anyway as long as it's consistent for the life of the DAGCircuit being exactly the same between multiple DAGCircuit objects has never been a guarantee (although we tried to improve it in #9569).~ This is addressed by: https://github.com/Qiskit/qiskit/pull/10729/commits/c57207d38a1e4fd38dc52829c20a920349903b46

### Details and comments

mtreinish committed:
Cache DAGOpNode sort keys

In a typical circuit there are only certain numbers of qargs that are
used regularly. The best examples are 1q gates on each qubit in the
target, or 2q gates on each edge in the connectivity graph. As the qargs
will be repeated many times in a typical circuit then for most
operations the sort key attribute is also not-unique (as it's just
composed of the positional indices of each qarg and carg). In very large
circuit compilation we spend a non-insignificant amount of time
constructing the sort key strings on nodes during the lifecycle of the
DAGCircuit as it gets transformed. This commit caches the sort keys so
that they're not rebuilt for every new DAGNode. This does add some extra
overhead to the construction the first time a DAGNode is created because
it has to do an additional qarg tuple construction and a dict lookup.
But this is worth it because in most circutis the cache hit rate will be
very high and we'll avoid the more expensive string construction for
most gates in a circuit.

There is also a potential shared state issue here if any qubit or clbit
objects are present in >1 DAGCircuit and the relative positions of those
bits differ between the circuits the sort key will not be the same as
before because the earlier qubit combinations. This is because the cache
is stored as a class attribute on the DAGOpNode class and will be reused
between circuits. In practice this is unlikely to come up very frequently.
But also it's impact is mitigated by these keys just being used for
sorting and the sorting hasn't always been 100% deterministic anyway as
long as it's consistent for the life of the DAGCircuit being exactly the
same between multiple DAGCircuit objects has never been a guarantee
(although we tried to improve it in #9569).

mtreinish committed:
Merge branch 'main' into cache-dag-sort-key

jakelishman commented on qiskit/dagcircuit/dagnode.py:
```diff
@@ -251,6 +251,7 @@ class DAGOpNode(DAGNode):
     """Object to represent an Instruction at a node in the DAGCircuit."""
 
     __slots__ = ["op", "qargs", "cargs", "sort_key"]
+    _key_cache = {}
```
I'm not a huge fan of us encoding this as static storage on the type object, since keys never get garbage collected within the lifecycle of the interpreter.  That said, since they're just integers, it's not necessarily awful, but I'm just super nervous about append-only memory usage.

mtreinish replied:
That's fair, I wasn't super worried about it. But with your suggestion below that will solve this too by moving the cache to be scoped to the dag instance.

jakelishman replied:
If we scope it by DAG instance, let's make `DAGCircuit.copy` and `.copy_empty_like` both share the same underlying cache dictionary; we don't even need to copy it, just add new references.

jakelishman commented on qiskit/dagcircuit/dagnode.py:
```diff
@@ -259,9 +260,13 @@ def __init__(self, op, qargs: Iterable[Qubit] = (), cargs: Iterable[Clbit] = (),
+                self.sort_key = key
+            else:
+                self.sort_key = ",".join(f"{dag.find_bit(q).index:04d}" for q in qargs)
+                self._key_cache[qargs] = self.sort_key
```
I don't think this cache lookup works; the same qubit instances could have different indices between two DAGs, but all the caching is shared per interpreter.  Switching to having this be an instance-level cache on the DAG would solve that, and let them be garbage collected as well.

mtreinish replied:
That works for me, I put a paragraph in the commit message about how I didn't think the global cache really mattered as long as the sort was consistent within a dag. But since we get the dag in `__init__` anyway we can just move it there and side step that potential issue.

jakelishman replied:
Also, I think we might want to see if we can work out a scheme where we do the key lookup based on the `qargs` and `cargs` tuples being shared separately - I know certainly for strings, the hash lookup is faster if you're using the referentially identical object to do the lookup, and we're moving towards a place where the `qargs` tuples _will_ be shared, and the `cargs` ones most often are already (since they're the empty tuple, which is a singleton).  I don't know for sure if tuples have the same behaviour, but at any rate, the `tuple(itertools.chain(...))` call is annoyingly slow if we can organise our cache to avoid it.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/10729/commits/c57207d38a1e4fd38dc52829c20a920349903b46

mtreinish committed:
Move sort key cache to be DAG instance scoped

This commit reworks the storage location for the cached dag sort keys.
In the previous commit the sort keys were stored as a class attribute on
the DAGOpNode. As was mentioned in that commit message this had
potential shared state issues. To address that potential issue this
commit moves the cache to be scoped to an instance of DAG as that is
really what the sort key is scoped to. At the same time the lookup key
for the cache is changed to be (qargs, cargs) instead of the flattened
tuple.

mtreinish committed:
Use a more accurate not overloaded variable name for the cache lookup key

mtreinish committed:
Merge branch 'main' into cache-dag-sort-key

jakelishman reviewed (approved):
Looks good to me.  Do you want to run any little benchmarks for the PR context?

mtreinish commented:
Sure I'll run some benchmarks now. I quickly ran some quick asv benchmarks when I first pushed this up and there wasn't a statistically significant change for the subset I ran so I'll expand the subset I run and post the results here.

mtreinish commented:
I ran asv locally and got these results:

```
Benchmarks that have improved:

       before           after         ratio
     [734e0618]       [dcc9a9dc]
     <main>       <cache-dag-sort-key>
-      78.8±0.8ms         71.6±1ms     0.91  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(0, 'sabre', 'dense')

Benchmarks that have stayed the same:

       before           after         ratio
     [734e0618]       [dcc9a9dc]
     <main>       <cache-dag-sort-key>
         243±20ms         279±10ms    ~1.15  queko.QUEKOTranspilerBench.time_transpile_bss(1, 'sabre')
       53.8±0.6ms         58.5±2ms     1.09  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(0, 'sabre', 'sabre')
          195±5ms          210±9ms     1.08  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(1, 'stochastic', 'sabre')
          269±3ms         287±10ms     1.07  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(1, 'stochastic', 'sabre')
       15.6±0.8ms         16.6±1ms     1.06  queko.QUEKOTranspilerBench.time_transpile_bigd(0, 'sabre')
          163±1ms         173±10ms     1.06  queko.QUEKOTranspilerBench.time_transpile_bss(0, 'sabre')
         26.5±2ms         28.0±3ms     1.06  queko.QUEKOTranspilerBench.time_transpile_bigd(2, 'sabre')
      2.36±0.01ms      2.48±0.05ms     1.05  converters.ConverterBenchmarks.time_circuit_to_dag(32, 8)
       18.1±0.1ms       19.1±0.5ms     1.05  converters.ConverterBenchmarks.time_circuit_to_dag(20, 128)
         83.3±3ms         87.7±6ms     1.05  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(1, 'sabre', 'sabre')
       77.2±0.3ms         81.0±6ms     1.05  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(2, 'sabre', 'noise_adaptive')
          137±3ms         143±10ms     1.05  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(1)
         76.8±4ms         79.9±5ms     1.04  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(2, 'sabre', 'dense')
       30.5±0.9ms       31.7±0.6ms     1.04  converters.ConverterBenchmarks.time_circuit_to_dag(32, 128)
          163±3ms          169±2ms     1.04  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(0)
          108±2ms          112±2ms     1.04  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(0, 'stochastic', 'sabre')
          429±2ms          444±4ms     1.04  mapping_passes.PassBenchmarks.time_sabre_swap(20, 1024)
       85.8±0.6ms         88.8±2ms     1.04  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(1, 'sabre', 'noise_adaptive')
          560±7ms         579±20ms     1.04  queko.QUEKOTranspilerBench.time_transpile_bss(2, 'sabre')
       4.15±0.01s       4.29±0.02s     1.03  qft.LargeQFTMappingTimeBench.time_sabre_swap(409, 'lookahead')
         77.6±2ms         79.9±4ms     1.03  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(1, 'sabre', 'dense')
       1.01±0.01s       1.04±0.01s     1.03  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(3, 'stochastic', 'sabre')
          278±6ms          285±5ms     1.03  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(0)
         1.03±0ms      1.06±0.01ms     1.03  converters.ConverterBenchmarks.time_circuit_to_dag(14, 8)
       2.05±0.02s       2.10±0.02s     1.02  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(2)
          479±3ms          490±5ms     1.02  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(3, 'sabre', 'dense')
       1.12±0.01s       1.15±0.01s     1.02  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(409, 10, 'lookahead')
       68.0±0.7ms       69.5±0.8ms     1.02  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 10, 'decay')
          4.14±0s       4.23±0.03s     1.02  qft.LargeQFTMappingTimeBench.time_sabre_swap(409, 'decay')
          258±3ms          263±3ms     1.02  qft.LargeQFTMappingTimeBench.time_sabre_swap(115, 'decay')
          588±3μs         600±10μs     1.02  converters.ConverterBenchmarks.time_circuit_to_dag(8, 8)
          259±4ms          264±2ms     1.02  qft.LargeQFTMappingTimeBench.time_sabre_swap(115, 'lookahead')
         613±10ms         625±10ms     1.02  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 100, 'lookahead')
      1.92±0.03ms      1.95±0.01ms     1.02  converters.ConverterBenchmarks.time_circuit_to_dag(2, 128)
          127±2ms          129±7ms     1.02  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(2, 'stochastic', 'sabre')
          239±2ms          244±5ms     1.02  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(3, 'sabre', 'sabre')
          100±2μs          102±2μs     1.02  converters.ConverterBenchmarks.time_circuit_to_dag(1, 8)
       17.0±0.1ms       17.3±0.2ms     1.02  converters.ConverterBenchmarks.time_circuit_to_dag(1, 2048)
       55.5±0.2ms       56.5±0.8ms     1.02  converters.ConverterBenchmarks.time_circuit_to_dag(53, 128)
        145±0.6ms          147±1ms     1.02  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(2, 'sabre', 'sabre')
       12.4±0.1ms       12.6±0.5ms     1.02  converters.ConverterBenchmarks.time_circuit_to_dag(14, 128)
       8.65±0.04s       8.80±0.07s     1.02  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(1081, 10, 'decay')
          181±6μs          184±1μs     1.02  converters.ConverterBenchmarks.time_circuit_to_dag(2, 8)
          469±3ms          477±5ms     1.02  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(3, 'sabre', 'sabre')
       73.4±0.4ms       74.7±0.7ms     1.02  converters.ConverterBenchmarks.time_circuit_to_dag(5, 2048)
          206±2ms          210±2ms     1.02  converters.ConverterBenchmarks.time_circuit_to_dag(14, 2048)
          617±8ms          628±9ms     1.02  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 100, 'decay')
         605±10ms         614±10ms     1.02  mapping_passes.PassBenchmarks.time_sabre_layout(5, 1024)
          333±2ms          338±3ms     1.02  transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(0)
          220±4ms          223±1ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(2)
       85.5±0.5ms       86.7±0.8ms     1.01  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(1, 'sabre', 'sabre')
         64.4±3ms         65.4±3ms     1.01  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 10, 'lookahead')
        133±0.6ms          135±1ms     1.01  converters.ConverterBenchmarks.time_circuit_to_dag(2, 8192)
          106±4ms         108±10ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(3)
        117±0.5ms        119±0.5ms     1.01  converters.ConverterBenchmarks.time_circuit_to_dag(8, 2048)
       3.26±0.01s       3.30±0.02s     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(3)
         933±30ms         944±10ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(1)
      4.20±0.09ms      4.25±0.06ms     1.01  converters.ConverterBenchmarks.time_circuit_to_dag(53, 8)
       3.02±0.01s       3.06±0.03s     1.01  mapping_passes.PassBenchmarks.time_sabre_layout(14, 1024)
       5.38±0.03s       5.44±0.01s     1.01  mapping_passes.PassBenchmarks.time_sabre_layout(20, 1024)
        300±0.9ms          303±1ms     1.01  converters.ConverterBenchmarks.time_circuit_to_dag(5, 8192)
          250±1ms          252±5ms     1.01  mapping_passes.PassBenchmarks.time_sabre_swap(14, 1024)
       72.4±0.4ms       72.9±0.5ms     1.01  converters.ConverterBenchmarks.time_circuit_to_dag(1, 8192)
       59.6±0.2ms         60.0±1ms     1.01  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(1, 'sabre', 'sabre')
        123±0.5ms          123±1ms     1.01  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(2, 'sabre', 'noise_adaptive')
          406±3ms          409±6ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(2)
         524±20ms          527±4ms     1.01  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(3, 'sabre', 'noise_adaptive')
       31.6±0.04s       31.8±0.04s     1.01  qft.LargeQFTMappingTimeBench.time_sabre_swap(1081, 'lookahead')
      4.38±0.04ms      4.40±0.08ms     1.01  converters.ConverterBenchmarks.time_circuit_to_dag(5, 128)
       3.16±0.01s       3.18±0.01s     1.01  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(409, 10, 'decay')
         69.6±1ms         70.0±2ms     1.01  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(0, 'sabre', 'sabre')
          32.3±0s        32.4±0.3s     1.01  qft.LargeQFTMappingTimeBench.time_sabre_swap(1081, 'decay')
         676±10ms          680±6ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(3, 'sabre', 'sabre')
          723±7ms         726±10ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(3)
          464±8ms          466±7ms     1.00  queko.QUEKOTranspilerBench.time_transpile_bntf(2, 'sabre')
       68.9±0.5ms       69.1±0.9ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(0)
          472±2ms          474±3ms     1.00  converters.ConverterBenchmarks.time_circuit_to_dag(8, 8192)
         56.4±1ms         56.6±1ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(0, 'sabre', 'noise_adaptive')
          618±5ms         620±10ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(3, 'sabre', 'dense')
          231±2ms          231±3ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(3, 'sabre', 'noise_adaptive')
       5.34±0.02s       5.35±0.02s     1.00  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(1081, 10, 'lookahead')
         323±10ms          324±4ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(3, 'stochastic', 'sabre')
          502±9ms         503±20ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(3, 'sabre', 'noise_adaptive')
      1.06±0.01ms      1.06±0.01ms     1.00  converters.ConverterBenchmarks.time_circuit_to_dag(1, 128)
          383±2μs          383±4μs     1.00  converters.ConverterBenchmarks.time_circuit_to_dag(5, 8)
          220±3ms          220±2ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(3, 'sabre', 'dense')
          274±2ms          273±2ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(1)
        134±0.8ms          134±3ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(2, 'sabre', 'noise_adaptive')
       1.48±0.01s       1.47±0.01s     1.00  queko.QUEKOTranspilerBench.time_transpile_bntf(3, 'sabre')
       1.55±0.02s       1.54±0.01s     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(0)
       7.08±0.1ms      7.04±0.07ms     1.00  converters.ConverterBenchmarks.time_circuit_to_dag(8, 128)
       83.2±0.6ms       82.8±0.6ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(1, 'sabre', 'dense')
       76.0±0.4ms       75.6±0.5ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(2, 'sabre', 'sabre')
         77.3±1ms         76.9±2ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(0, 'sabre', 'noise_adaptive')
         64.6±2ms         64.2±2ms     0.99  mapping_passes.PassBenchmarks.time_sabre_swap(5, 1024)
          247±7ms          246±9ms     0.99  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(1)
       72.2±0.7ms         71.6±2ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(0, 'sabre', 'noise_adaptive')
          249±4ms          247±5ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(0, 'stochastic', 'sabre')
       58.8±0.4ms       58.2±0.4ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(1, 'sabre', 'dense')
       60.8±0.7ms       60.1±0.7ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(1, 'sabre', 'noise_adaptive')
       98.3±0.6ms         97.0±2ms     0.99  queko.QUEKOTranspilerBench.time_transpile_bigd(3, 'sabre')
       33.0±0.9ms       32.5±0.6ms     0.98  converters.ConverterBenchmarks.time_circuit_to_dag(2, 2048)
          110±4ms          108±3ms     0.98  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(3)
          244±6ms          240±2ms     0.98  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(1)
       1.05±0.02s       1.03±0.02s     0.98  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(3, 'stochastic', 'sabre')
         205±10ms          201±2ms     0.98  queko.QUEKOTranspilerBench.time_transpile_bntf(1, 'sabre')
      1.52±0.03ms         1.49±0ms     0.98  converters.ConverterBenchmarks.time_circuit_to_dag(20, 8)
       81.9±0.8ms       80.5±0.5ms     0.98  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(1, 'sabre', 'noise_adaptive')
          222±3ms          217±2ms     0.98  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(2)
          267±4ms          261±3ms     0.98  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(2, 'stochastic', 'sabre')
       18.1±0.5ms       17.6±0.5ms     0.97  queko.QUEKOTranspilerBench.time_transpile_bigd(1, 'sabre')
       1.15±0.03s       1.11±0.01s     0.97  queko.QUEKOTranspilerBench.time_transpile_bss(3, 'sabre')
         77.1±4ms         73.5±1ms     0.95  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(0, 'sabre', 'dense')
         56.7±3ms       53.6±0.5ms     0.95  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(0, 'sabre', 'dense')
          194±9ms          183±4ms     0.94  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(0, 'stochastic', 'sabre')
         156±10ms          147±3ms     0.94  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(2, 'sabre', 'sabre')
         340±10ms          320±4ms     0.94  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(2, 'stochastic', 'sabre')
         77.6±4ms       72.6±0.4ms     0.94  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(0, 'sabre', 'sabre')
         169±10ms          158±2ms     0.94  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(2, 'sabre', 'dense')
         157±10ms          146±3ms     0.93  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(2, 'sabre', 'dense')
          125±6ms          114±2ms     0.91  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(1, 'stochastic', 'sabre')
         138±10ms        124±0.7ms    ~0.90  queko.QUEKOTranspilerBench.time_transpile_bntf(0, 'sabre')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

I also did some manual testing locally and on really large circuits from QASMbench with optimization level 0 it saves about 10% of the current runtime (which matches the results in asv). But for higher optimization levels the sort key time isn't significant relative to the other stuff going on so you don't really see it as much.
